autox/informer.go

```diff
@@ -3,154 +3,359 @@ package autox
+	}
+}
 
-	uObj := obj.(*unstructured.Unstructured)
-	resourceName := uObj.GetName()
-	// example label: iter8.tools/autox-group=hello
-	labels := pruneLabels(uObj.GetLabels())
+func updateObject(ns string, gvr schema.GroupVersionResource, autoXConfig config) func(oldObj, obj interface{}) {
+	return func(oldObj, obj interface{}) {
+		log.Logger.Debug("Update:", obj)
+		handle(obj, ns, gvr, autoXConfig)
+	}
+}
 
-	// check if the app name matches the name in the autox label
-	if autoxLabelName, ok := labels[autoxLabel]; ok && resourceName == autoxLabelName {
-		// Delete Helm charts
-		deleteHelmReleases(resourceName)
+func deleteObject(ns string, gvr schema.GroupVersionResource, autoXConfig config) func(obj interface{}) {
+	return func(obj interface{}) {
+		log.Logger.Debug("Delete:", obj)
+		handle(obj, ns, gvr, autoXConfig)
 	}
 }
 
+func addObjectHelper(uObj *unstructured.Unstructured, prunedLabels map[string]string, autoXConfig config) error {
+	return installHelmReleases(prunedLabels, uObj.GetNamespace(), autoXConfig)
+}
+
+func deleteObjectHelper(uObj *unstructured.Unstructured, prunedLabels map[string]string, autoXConfig config) error {
+	return deleteHelmReleases(prunedLabels, uObj.GetNamespace(), autoXConfig)
+}
+
 type iter8Watcher struct {
 	factories map[string]dynamicinformer.DynamicSharedInformerFactory
 }
 
-func newIter8Watcher() *iter8Watcher {
+func newIter8Watcher(autoXConfig config) *iter8Watcher {
 	w := &iter8Watcher{
 		factories: map[string]dynamicinformer.DynamicSharedInformerFactory{},
 	}
-	// for each namespace, resource type configure Informer
-	for _, ns := range iter8ResourceConfig.Namespaces {
+
+	// aggregate all triggers (namespaces and GVR) from the releaseGroupConfig
+	// triggers map has namespace as its key and the object GVRs within the namespace that it is watching as its value
+	triggers := map[string][]schema.GroupVersionResource{}
+	for _, releaseGroupSpec := range autoXConfig.Specs {
+
+		namespace := releaseGroupSpec.Trigger.Namespace
+		gvr := schema.GroupVersionResource{
+			Group:    releaseGroupSpec.Trigger.Group,
+			Version:  releaseGroupSpec.Trigger.Version,
+			Resource: releaseGroupSpec.Trigger.Resource,
+		}
+
+		// add namespace and GVR to triggers
+		triggers[namespace] = append(triggers[namespace], gvr)
+	}
+
+	// for each namespace and resource type, configure an informer
+	for ns, gvrs := range triggers {
 		w.factories[ns] = dynamicinformer.NewFilteredDynamicSharedInformerFactory(k8sClient.dynamicClient, 0, ns, nil)
-		for _, gvr := range iter8ResourceConfig.Resources {
+
+		// TBD: optimize the informers by supplying the trigger object name, perhaps through list options
+		for _, gvr := range gvrs {
 			informer := w.factories[ns].ForResource(gvr)
 			informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-				AddFunc:    addObject,
-				UpdateFunc: updateObject,
-				DeleteFunc: deleteObject,
+				AddFunc:    addObject(ns, gvr, autoXConfig),
```

Comment:
I feel these func signatures still need some fine tuning, based on our slack discussion. Could we please discuss when we meet during one of our dev meetings?

***

The logic is good, but it can be made a bit more robust, simply by watching for each trigger resource. FieldSelectors might help with this.